src/DataStructures/Mesh.hpp

```diff
@@ -0,0 +1,87 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class template Mesh.
+
+#pragma once
+
+#include "DataStructures/Index.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+
+template <size_t Dim>
+class Mesh : public Index<Dim> {
+ public:
+  explicit Mesh(
+      const size_t uniform_extents = std::numeric_limits<size_t>::max(),
```

Comment:
remove uniform_ from the input names and remove the default arguments

Reply:
I put the defaults here since LGL is actually the default mesh used everywhere in the code. So when a `Mesh` is required where people don't really care about the choice of basis, LGL is a reasonable choice. All functionality should work for any basis anyway, so it's OK if people don't care about this too much until it becomes necessary. So e.g. in tests where other functionality is tested, such as making sure some compute items work, this should be independent of the choice of basis and the defaults can be used. Then to explicitly make sure some functionality actually works on mixed meshes, this is where the choice of bases is specified.

I also set it up this way to further decouple the low-level `Spectral` namespace from code that uses the `Mesh`: Usually code only needs to work with the `Mesh` and doesn't even need to include `Spectral.hpp`. Only when it becomes necessary to work with a specific basis and such does it have to use the `Spectral` namespace.

Do you agree with this architecture, or have another opinion? We can discuss in the hangout, too.

Reply:
We want to be able to forward declare the enum class, so that DataStructures isn't tied to NumericalAlgorithms too tightly

Reply:
Hmm I see, good point. Since the `Mesh` is all about placing the `Spectral` algorithms into context, we could move it from `DataStructures` to `Spectral`. Then we could keep the defaults and also the member functions. Or is there another way to keep the defaults?

Reply:
I've thought about that a little bit, but I'm not sure I see having to specify the basis and quadrature as a bad thing, actually. It adds code but also removes any assumptions

Reply:
Actually objects such as the differentiation matrix are obtained from the mesh that is passed alongside every `DataVector`, so it should always be correct (A `DataVector` that represents field values has no meaning without a `Mesh` anyway). But I agree that what you proposed is better, refactoring now.

Only one issue: If I am not mistaken we need a default constructor for `Mesh` that takes no arguments, so how can we set this up?

Reply:
Okay, great :)

I tested this yesterday to make sure everything would work before we suggested the change. You can just use `Mesh() noexcept = default;` and then do `std::array<Basis, Dim> basis_{};`